### PR TITLE
Blogs page

### PR DESCRIPTION
**Pull request by mohammad-jundi**
Connected blogs to WP API 
Style added 
for the Search bar function, I will do it in a separate PR

### CODE REVIEW

**Comment by mohammad-jundi** [comment]
image

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: consts with uppercase should be outside of function just below the import

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: please delete console.log here and on line 21

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: maybe smth like BLOGS_URL could be more descriptive?

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: typo in information. also, you should add translation

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: typo in positively

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: Blog image

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: i would say this should be blogImg since you map and create blog one by one. same with blogTitle

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: img is a self-closing tag

File: src/components/Blogs/index.scss
Comment (jennetturkkan) [inline src/components/Blogs/index.scss]: i didn't see this col in jsx file. where does it come from?

File: src/components/Blogs/index.scss
Comment (jennetturkkan) [inline src/components/Blogs/index.scss:21]: i wonder if this position absolute will create any issues in future? i had to delete it from seekHelp section in homepage because it was rendering not the way we wanted

**Review by jennetturkkan — CHANGES_REQUESTED** [review]

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx:40]: is there other way of adding shadow effect? it looks like this div can be removed here

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx]: I think class name is a bit confusing since you have already bootstrap container above this div

File: src/components/Blogs/index.scss
Comment (jennetturkkan) [inline src/components/Blogs/index.scss]: if this is a bootstrap class name, you should add comment above it `// Bootstrap class`

File: src/components/Blogs/index.jsx
Comment (louisrli) [reply src/components/Blogs/index.jsx]: these constants dont go in fn body, move under imports after line 4

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: so imagine you are me. what is blogs? all i know is blogs is an empty array. add a comment please abovei t

also we usually use "blogPosts" to refer to the posts and blog to refer to the whole site specifically.

File: src/components/Blogs/index.scss
Comment (louisrli) [reply src/components/Blogs/index.scss:21]: absolute position should generally be avoided where possible because it makes code hard to change

**Review by louisrli — COMMENTED** [review]
+jennets comments

File: src/components/Blogs/index.scss
Comment (mohammad-jundi) [reply src/components/Blogs/index.scss:21]: I used it inside nesting to avoid any kind of conflict in the future. tried different things, absolute was the most accurate one to use. I'll search if there is something else I can use

File: src/components/Blogs/index.scss
Comment (mohammad-jundi) [reply src/components/Blogs/index.scss]: these from bootstrap, I will comments if that's allowed, will ask Ammar about it.

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: this name caused a problem when running the tests so I had to change it to this name since the tag is already for images

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: it's a container that contains the blogs container, if you have any suggestions please let me know them, I can change it to blogsSection if it's much more convenient.

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.jsx]: in alt? i don't think that `alt="Blog image"` will create some problem

File: src/components/Blogs/index.scss
Comment (jennetturkkan) [reply src/components/Blogs/index.scss]: but you are not using it anywhere. it's not even imported in index.jsx file

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.jsx]: yeah, it's more clear. thank you

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: fixed

File: src/components/Blogs/index.scss
Comment (mohammad-jundi) [reply src/components/Blogs/index.scss]: done

File: src/components/Blogs/index.scss
Comment (jennetturkkan) [reply src/components/Blogs/index.scss:21]: my concern is that it will create a lot of problems with responsiveness. but it's your call in any case.

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx:40]: tried to use box-shadow, didn't work for me. this one worked

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: done

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: fixed

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: fixed

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: image

Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: after changing the name in alt, the test passes

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: it makes much more sense, thanks for the suggestion

File: src/components/Blogs/index.scss
Comment (mohammad-jundi) [reply src/components/Blogs/index.scss]: it comes with the package I guess, I am not sure. however when I inspected it and tried to style it worked.

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.jsx]: you can use blog.title.rendered as an alt. it won't cause error and will be unique for every blog post.

**Comment by mohammad-jundi** [comment]
PTAL @louisrli

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: newline

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: BLOG_API_URL

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx:12]: consider a newline above here too. it's like a "paragraph" in an essay. you're doing something new here so you separate it from the thing above.

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: this should be shownBlogPostsArr. blogPosts is actually the fetchedBlogPosts, but you are only showing N of them.

but actually, you shouldn't do it like this. you should use `per_page` in the API. even without knowing the wordpress API it's possible to know that there's probably something like this because usually APIs dont make you load EVERYTHING then cut it in your code, they make you say exactly how much you want to load beforehand (this saves them network)

see per_page

https://developer.wordpress.org/rest-api/reference/posts/#arguments

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: translations?

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: don't manually line break like this. let the CSS take care of it. it also doesn't play well with translation

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: I read this documentation before start with WP API and wasn't so clear to me. 
Thanks for referring it to me again it helped me a lot in cleaning the code. Maybe it's not the cleanest but I am sure it's much better now

**Comment by mohammad-jundi** [comment]
PTAL @louisrli 
fixed the extra `&nbsp;` that was showing in titles. 
I am not sure if I did the translation part in the right way, 
I will add the search function in Containers/BlogPage, please advise if it's fine in this way 
image

File: src/containers/BlogPage/index.jsx
Comment (jennetturkkan) [inline src/containers/BlogPage/index.jsx]: this is not right. you can check how we did translation before. you need to add this text to translation.json and then select it here.

File: src/components/Blogs/index.jsx
Comment (jennetturkkan) [inline src/components/Blogs/index.jsx:20]: add new line after this

File: src/containers/BlogPage/index.jsx
Comment (jennetturkkan) [inline src/containers/BlogPage/index.jsx]: you need to add translation here too

File: src/containers/BlogPage/index.scss
Comment (jennetturkkan) [inline src/containers/BlogPage/index.scss]: this can be padding: 0 30%

File: src/containers/BlogPage/index.scss
Comment (jennetturkkan) [inline src/containers/BlogPage/index.scss]: is it a font-size?

**Review by jennetturkkan — CHANGES_REQUESTED** [review]

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx:5]:
```
const POSTS_PER_PAGE = 9
const BLOG_API_URL = `blahblah?per_page=${POSTS_PER_PAGE}`
```

File: src/containers/BlogPage/index.scss
Comment (louisrli) [inline src/containers/BlogPage/index.scss]: space before B

File: src/containers/BlogPage/index.jsx
Comment (mohammad-jundi) [reply src/containers/BlogPage/index.jsx]: I wanted to do it in this way, but I wasn't sure if this will cause a conflict in the JSON file since Hiba was adding new data. So I kept it in this way till she finishes adding the data so that I can add to it later.

Comment (louisrli) [reply src/containers/BlogPage/index.jsx]: just write to en.json first.

**Comment by mohammad-jundi** [comment]
PTAL @louisrli

File: src/components/Blogs/__snapshots__/index.test.jsx.snap
Comment (louisrli) [inline src/components/Blogs/__snapshots__/index.test.jsx.snap]: this test should show at least one mock blog post. you will have to mock fetch so that it returns some blog post data. google "mock fetch jest unit test" or work with ammar in this one. but it will be worth your time to do this, testing is by far the most popular interview question I've seen from companies so consider yourself lucky to get the chance to do this one😀

File: src/components/Blogs/index.jsx
Comment (louisrli) [inline src/components/Blogs/index.jsx]: hmm I'm confused. this isn't coming from the translations file. work with hiba or jennet here to solve this, I'm not sure what this page is supposed to be so I can't advise

**Review by louisrli — CHANGES_REQUESTED** [review]

File: src/components/Blogs/__snapshots__/index.test.jsx.snap
Comment (mohammad-jundi) [reply src/components/Blogs/__snapshots__/index.test.jsx.snap]: I always thought I am the lucky one, now I am sure that I am :)

File: src/components/Blogs/index.jsx
Comment (mohammad-jundi) [reply src/components/Blogs/index.jsx]: You are right this is not from the translations file, this one is being fetched from WP API to show only the titles for the posts. I had to use `dangerouslySetInnerHTML` because the data was fetched with extra `&nbsp;` that wasn't supposed to be shown.

Comment (louisrli) [reply src/components/Blogs/index.jsx]: right but it has a t() around it -- i think it's not supposed to

**Comment by mohammad-jundi** [comment]
@louisrli 
Added the search function, is it ok if the test updated after merging,? if you don't have any other comments of course

**Comment by louisrli** [comment]
> @louisrli
> Added the search function, is it ok if the test updated after merging,? if you don't have any other comments of course

no, because i already know that means the test will not get added. and if this were work at a company, you'd have to figure it out anyway :)

**Comment by mohammad-jundi** [comment]
@louisrli PTAL

**Review by louisrli — CHANGES_REQUESTED** [review]
why are 13000 lines changed

**Comment by mohammad-jundi** [comment]
> why are 13000 lines changed

sometimes vsc says the files are changing but actually they are not, and the this problem solution is to commit them even though there is no change at all. This time changed 13 000 line !!! I will try to go back to the previous commits and fix. 

update: 
after checking the files, it seems that yarn.lock file have double files in it (I don't know how or why). I am working on resolving it.

**Comment by hiba-machfej** [comment]
I noticed when opening the branch that carousels on the home page were not there anymore. I noticed that global classes were used like .container or .col which was affecting the home page. I just removed the scss related to container and col and added Row from bootstrap and styling of the home page is working now.  this is how it was appearing before :
Screen Shot 2020-11-19 at 1 39 40 PM
Screen Shot 2020-11-19 at 1 39 53 PM

File: public/locales/en/translation.json
Comment (louisrli) [inline public/locales/en/translation.json]: } on new line

File: src/components/Blogs/__snapshots__/index.test.jsx.snap
Comment (louisrli) [inline src/components/Blogs/__snapshots__/index.test.jsx.snap]: this does not seem right, it is undefined

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx]: this is what your test is doing

1. create a random new function called Blogs.fetchBlogPosts that does nothing (this is not the same as the function in your thing)
2. call this function. console.log title here. it's undefined
3. compare this "undefined" title to a snapshot that also says "undefined."

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx]:
```
   const fetchedBlogPosts = await data.json();
    setBlogPosts(fetchedBlogPosts);
```

see this line in your code

you want what you return on line 9-11 to be valid data so that it makes your snapshot test show a blog

please ask for another review when the snapshot test shows an example of a blog post (i can tell that you're not even checking your snapshot test because of what it says right now, otherwise you'd know that the test you wrote doesn't work)

the idea here is that you are mocking fetch so that it gives you fake wordpress data which then creates an example of a blog post in the snapshot test

work with your teammates or ammar to get this finished today please. @jennethydyrova @Jamil-Bailony @hiba-machfej @Ammar-64

**Review by louisrli — CHANGES_REQUESTED** [review]

**Comment by mohammad-jundi** [comment]
> why are 13000 lines changed

I was checking the files, regarding the changed lines. they are from the packages that installed recently, when running yarn install to update the branch with dev it deleted the old package and downloaded the new one which caused the changes in 13000 lines. I am afraid if I remove the file and download it again it will still give me 13000. please advise if it's normal thing to have @louisrli

**Comment by mohammad-jundi** [comment]
> I noticed when opening the branch that carousels on the home page were not there anymore. I noticed that global classes were used like .container or .col which was affecting the home page. I just removed the scss related to container and col and added Row from bootstrap and styling of the home page is working now. this is how it was appearing before :
> Screen Shot 2020-11-19 at 1 39 40 PM
> Screen Shot 2020-11-19 at 1 39 53 PM

Thanks it looks much more better now

**Comment by jennetturkkan** [comment]
@louisrli I wrote the test, but I am not sure if it want you wanted. can you please check?

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx:31]: it seems like it works but you only need one test, and that's the snapshot test, and so this rendering should be moved into the snapshot test

also see:

Comment (louisrli) [inline src/components/Blogs/index.test.jsx]: oops wrote this twice

it seems like it works but you only need one test, and that's the snapshot test, so delete this one and render the snapshot so that the snapshot test shows a blog post please

https://reactjs.org/docs/testing-recipes.html#data-fetching

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx]: I moved test. can you check?

Comment (louisrli) [reply src/components/Blogs/index.test.jsx]: but what about the updated snapshot

**Comment by louisrli** [comment]
the images seem squahsed in the screenshot jundi put too

File: src/components/Blogs/index.test.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx]: blogs are not being shown in snapshot, because I think useEffect doesn't run in snapshot

**Comment by jennetturkkan** [comment]
> the images seem squahsed in the screenshot jundi put too

styling will be updated

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [reply src/components/Blogs/index.test.jsx]: well, you have two calls to `<blogs>` right now. perhaps try having one.

**Review by louisrli — COMMENTED** [review]
i suggest reading the doc i linked and thinking how it can be applied to this code

**Comment by louisrli** [comment]
the only difference should be what you assert at the end of the test (they do some "is it in the document stuff", you would do a snapshot)

**Comment by jennetturkkan** [comment]
@louisrli test works. can you check?

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx:8]: what happens if you delete lines 10 through 18? and line 8

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx]: and just use render(<Blogs/>)

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx]: it doesn't work without ReactDOM.render(). I tried with render()

File: src/components/Blogs/index.test.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx:8]: if I delete, what can I compare snapshot with?

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [reply src/components/Blogs/index.test.jsx]: sry maybe not render. i don't remember but just follow the normal format for snapshots.

```
const tree = renderer.create(<MemberCard />).toJSON();
  expect(tree).toMatchSnapshot();
```

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx]: a bunch of these imports are unused too i believe

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx]: yes, i will delete them now

File: src/components/Blogs/index.test.jsx
Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx]: it turns out I was importing the wrong render. Now I changed it and it works.

**Comment by jennetturkkan** [comment]
Fixed images and header styling, added some margin, added row

Screen Shot 2020-11-20 at 14 50 18

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [inline src/components/Blogs/index.test.jsx:38]: the thing is, the example code i showed you doesn't use render. it uses render**er**
```
const tree = renderer.create(<MemberCard />).toJSON();
  expect(tree).toMatchSnapshot();
```

basically, i'm fairly sure you don't need to do this whole container thing. just literally copy the code above and put it in the async, change it to Blogs.

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx:38]: like this? 
```describe("Blogs", () => {
  it("Renders without crashing", async () => {
    const mockData = [
      {
        title: {
          rendered: "You can find safe place for yourself here",
        },
        jetpack_featured_media_url:
          "https://safeplace102505649.files.wordpress.com/2020/11/womentogether.jpg",
        id: 78,
      },
    ];

    jest.spyOn(global, "fetch").mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(mockData),
      })
    );

    await act(async () => {
      const tree = renderer.create(<Blogs />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
```

**Comment by louisrli** [comment]
try it, does it work? :)

**Comment by jennetturkkan** [comment]
I tried it before and it doesn't. that's why I added this container thing. 
Screen Shot 2020-11-20 at 16 04 10

File: src/components/Blogs/index.test.jsx
Comment (louisrli) [reply src/components/Blogs/index.test.jsx:38]:
```
      expect(tree).toMatchSnapshot();
```

this needs to be outside the act i believe

you are basically awaiting the fetch using act

and THEN you want to match the snapshot

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx:38]: then tree will be undefined or should i declare it outside of function?

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx:38]: I tried this before too. It doesn't work
``` 
let tree;
    await act(async () => {
      tree = renderer.create(<Blogs />).toJSON();
    });
    expect(tree).toMatchSnapshot(); 
```

Comment (jennetturkkan) [reply src/components/Blogs/index.test.jsx:38]: this is what I get
Screen Shot 2020-11-20 at 16 12 14

**Review by louisrli — APPROVED** [review]
ok thats fine then

**Pull request merged**
